File: sdks/node/src/middleware.ts
import type { OutcomeEvent } from "./events";
import type { EnqueueOptions } from "./types";

/** Context for a task as it executes on a worker. */
export interface TaskContext {
  jobId: string;
  taskName: string;
  args: unknown[];
}

/**
 * Context for a job being enqueued, passed to {@link Middleware.onEnqueue} before
 * serialization. Mutate `args`/`options` in place to validate, redact, or rewrite
 * the job; throw to abort the enqueue.
 */
export interface EnqueueContext {
  readonly taskName: string;
  /** Positional args, mutable before they are serialized. */
  args: unknown[];
  /** Enqueue options, mutable before they reach the core. */
  options: EnqueueOptions;
}

/**
 * Cross-cutting hooks around task execution and job outcomes. Register with
 * {@link Queue.use}. `onEnqueue` runs (sync) on the enqueuing side before
 * serialization; `before`/`after`/`onError`/`onSleep` wrap execution (awaited,
 * counted toward the timeout); the outcome hooks fire after the core decides
 * the result.
 */
export interface Middleware {
  /**
   * Stable name used by the dashboard's per-task middleware toggles.
   * Defaults to the class name for class-based middleware.
   */
  name?: string;
  onEnqueue?(ctx: EnqueueContext): void;
  before?(ctx: TaskContext): void | Promise<void>;
  after?(ctx: TaskContext, result: unknown): void | Promise<void>;
  /**
   * Called when an attempt ends in a `ctx.step.sleep`. Pairs with `before`.
   *
   * Every `before` is matched by exactly one of `after` or `onSleep`. A sleep
   * is not a result: `after(ctx, undefined)` is indistinguishable from "the
   * task returned undefined", which would close a tracing span as a success and
   * increment a success counter — both wrong for an attempt that has not
   * finished. Middleware that opens something in `before` and implements only
   * `after` leaks it on a sleep, and the runner warns once about it.
   *
   * @param wakeAt Deadline the job was rescheduled to, in Unix milliseconds.
   */
  onSleep?(ctx: TaskContext, wakeAt: number): void | Promise<void>;
  onError?(ctx: TaskContext, error: unknown): void | Promise<void>;
  onCompleted?(event: OutcomeEvent): void;
  onRetry?(event: OutcomeEvent): void;
  onDeadLetter?(event: OutcomeEvent): void;
  onCancel?(event: OutcomeEvent): void;
}

File: sdks/java/src/main/java/org/byteveda/flexiq/middleware/Middleware.java
package org.byteveda.flexiq.middleware;

import org.byteveda.flexiq.events.OutcomeEvent;

/**
 * Cross-cutting hooks around enqueue, task execution, and job outcomes. All are
 * optional. {@code onEnqueue} runs on the producer before serialization;
 * {@code before}/{@code after}/{@code onError} wrap execution; the outcome hooks
 * fire after the core decides the result. Register with {@link org.byteveda.flexiq.FlexiQ#use}.
 *
 * <p><b>Every {@code before} is matched by exactly one of {@code after} or
 * {@code onSleep}.</b> An attempt that ended in a durable {@code step.sleep} has
 * not finished, so it gets {@code onSleep}; anything that opens state in
 * {@code before} and implements only {@code after} leaks it whenever a task
 * sleeps, and the worker warns once when it sees that shape.
 */
public interface Middleware {
    /**
     * On the producer, before the payload is serialized.
     *
     * @param context the job being enqueued; rewrite its payload or options here,
     *     or throw to abort the enqueue
     */
    default void onEnqueue(EnqueueContext context) {}

    /**
     * On the worker, before the handler runs.
     *
     * @param context this execution, whose {@code attributes()} carry state to the
     *     matching {@code after} or {@code onSleep}
     */
    default void before(TaskContext context) {}

    /**
     * On the worker, after the handler returned.
     *
     * @param context this execution, the same instance {@code before} saw
     * @param result what the handler returned; {@code null} for a void handler
     */
    default void after(TaskContext context, Object result) {}

    /**
     * On the worker, after the handler threw.
     *
     * @param context this execution, the same instance {@code before} saw
     * @param error what the handler threw, before the core classifies it
     */
    default void onError(TaskContext context, Throwable error) {}

    /**
     * The attempt ended in a durable {@code step.sleep}: no result, no failure.
     *
     * <p>The job is {@code Pending} at {@code wakeAt} and will run again from
     * its memoized steps, so close whatever {@code before} opened rather than
     * completing it. {@code after(context, null)} would read as "the task
     * returned null" — a success to a timer, a span or a counter — which is
     * exactly what this hook exists to avoid.
     *
     * @param context this execution, the same instance {@code before} saw
     * @param wakeAt the deadline the job was rescheduled to, in Unix milliseconds
     */
    default void onSleep(TaskContext context, long wakeAt) {}

    /**
     * The job finished successfully.
     *
     * @param event the outcome the core recorded
     */
    default void onCompleted(OutcomeEvent event) {}

    /**
     * The attempt failed and the job will be tried again.
     *
     * @param event the outcome the core recorded, carrying the attempt's error
     */
    default void onRetry(OutcomeEvent event) {}

    /**
     * The job exhausted its retries and was moved to the dead-letter queue.
     *
     * @param event the outcome the core recorded, carrying the final error
     */
    default void onDeadLetter(OutcomeEvent event) {}

    /**
     * The job was cancelled rather than run to a verdict.
     *
     * @param event the outcome the core recorded
     */
    default void onCancel(OutcomeEvent event) {}
}
